server/src/core/server/models/notifications/notification.ts

```diff
@@ -0,0 +1,110 @@
+import { MongoContext } from "coral-server/data/context";
+
+import { ConnectionInput, Query, resolveConnection } from "../helpers";
+import { TenantResource } from "../tenant";
+import { User } from "../user";
+
+export interface Notification extends TenantResource {
+  readonly id: string;
+
+  readonly tenantID: string;
+
+  createdAt: Date;
+
+  ownerID: string;
+
+  reportID?: string;
+  commentID?: string;
+
+  title?: string;
+  body?: string;
+}
+
+type BaseConnectionInput = ConnectionInput<Notification>;
+
+export interface NotificationsConnectionInput extends BaseConnectionInput {
+  ownerID: string;
+}
+
+export const retrieveNotificationsConnection = async (
+  mongo: MongoContext,
+  tenantID: string,
+  input: NotificationsConnectionInput
+) => {
+  const query = new Query(mongo.notifications()).where({
+    tenantID,
+    ownerID: input.ownerID,
+  });
+
+  query.orderBy({ createdAt: -1 });
+
+  if (input.after) {
+    query.where({ createdAt: { $lt: input.after as Date } });
+  }
+
+  return resolveConnection(query, input, (n) => n.createdAt);
+};
+
+export const createNotification = async (
+  mongo: MongoContext,
+  notification: Notification
+) => {
+  const op = await mongo.notifications().insertOne(notification);
+
+  return op.result.ok ? notification : null;
+};
+
+interface LastSeenNotificationChange {
+  $set: {
+    lastSeenNotificationDate?: Date | null;
+  };
+}
+
+export const markLastSeenNotification = async (
+  tenantID: string,
+  mongo: MongoContext,
+  user: Readonly<User>,
+  notificationDates: Date[]
+) => {
+  if (!notificationDates || notificationDates.length === 0) {
+    return;
+  }
+
+  let max = new Date(0);
+  for (const date of notificationDates) {
+    if (max.getTime() < date.getTime()) {
+      max = date;
+    }
+  }
+
+  const change: LastSeenNotificationChange = {
+    $set: { lastSeenNotificationDate: user.lastSeenNotificationDate },
+  };
+
+  if (
```

Comment:
This works as is just fine, just a small (subjective) suggestion for this check for whether the `lastSeenNotificationDate` should be set to `max` and the one below as well. It may make it easier to read at a glance what's going on here if it was something more like:
```
const newNotificationSinceLastSeen = user.lastSeenNotificationDate && user.lastSeenNotificationDate.getTime() < max.getTime();
const notificationsNeverViewed = user.lastSeenNotificationDate === null || user.lastSeenNotificationDate === undefined;
if (newNotificationSinceLastSeen || notificationsNeverViewed) {
change.$set.lastSeenNotificationDate = max;
}
```

Reply:
I'm going to have to think about this, my brain is finding the above rather confusing, I do like your naming of the conditions being checked, that's cool. I'll have a think and do what I can to make this code more read-able!

Reply:
Fair enough, sounds good, ty...mostly that both the `if` and the `else if` led to the same outcome seemed like something unexpected to me when I was following along with the code

Reply:
refactored, I think it will make more sense now